### PR TITLE
Port Polykey.io Theme to Docs

### LINKED ISSUE #6: Port Polykey.io Theme to polykey.io/docs
### Specification

The Polykey.io theme should be brought into the documentation site. The main `polykey.io` is located in https://gitlab.com/MatrixAI/Engineering/Polykey/polykey.io.

The current documentation site is based off Docusaurus classic theme. https://docusaurus.io/docs/api/themes/@docusaurus/theme-classic. This means the current docs site looks very similar to the official docusaurus site and to many other docusaurus documentation websites.

Docusaurus classic theme itself uses a "framework" called Infima https://infima.dev/. This means all of component structure is based on Infima.

We want to preserve this classic theme as much as possible, while customising the theme/overriding parts of the theme to fit the Polykey.io design elements.

In order to do this, Docusaurus provides 3 different ways of changing the existing theme:

1. Override the CSS with a global CSS file.
2. Wrapping an existing React component with a custom React component, which is loaded ahead in priority.
3. Replacing an existing React component with a custom React component, which is loaded ahead in priority.

For overriding the CSS, see: https://docusaurus.io/docs/styling-layout.

It will require a new `./src/css/custom.css` file, which is loaded into the `./docusaurus.config.js` file.

The CSS file can specify theme class names, infima class names, or CSS module class names. Prefer using the theme class names, then the infima class names before creating custom CSS module class names.

When it comes to overriding the React components prefer wrapping before replacing as this is more stable in the long term. I have already made an example of overriding an existing React component. This is located in:

```
»» ~/Projects/Polykey-Docs
 ♖ tree src                                                                                                                              (master) pts/2 16:20:06
src
└── theme
    └── Navbar
        └── index.tsx

2 directories, 1 file
```

This was produced through the command:

```sh
npm run swizzle @docusaurus/theme-classic NavBar -- --wrap --typescript
```

This will create a new `./src/theme/Navbar/index.tsx` file. This file is loaded ahead of the original `./node_modules/@docusaurus/theme-classic/src/theme/Navbar/index.tsx` file. This means that the new file will be loaded instead of the original file.

Docusaurus calls this overriding concept "swizzling". This is explained here: https://docusaurus.io/docs/swizzling.

Take note of the all the components available to be swizzled with `npm run swizzle -- --list`.

#### Navigation Bar

image

image

The navigation bar in `polykey.io/docs` must be made to look **exactly** the same as the one in `polykey.io`. This means the `Docs` link must go to `polykey.io/docs`. The other links on the navigation bar must open within the same frame (target must be `_self`). This ensures that the docs looks like just another subpage as part of the main website.

The docs navbar currently has 2 additional things that should be removed:

image

The `GitHub` link should be unnecessary. We already have a GitHub icon on the navbar on https://polykey.io/.

> Note that I have already removed the dark mode button.

During mobile usage, notice that the navbar collapses.

image

image

This collapse looks different from the current `polykey.io` website. On the `polykey.io` website, the menu is put on the right-hand side.

The menu for docusaurus is superior to the menu currently on `polykey.io`. However some changes should be made.

The menu button should be moved to the right-hand side to be similar to `polykey.io`.

image

However leave the menu contents the same. We will only address this once we start migrating `polykey.io` to docusaurus.

The links here are specified in `./docusaurus.config.js` under `config.themeConfig.navbar.items`.

Ensure that hover colours are also the same.

Ensure that the email configuration is the same. It should be subscribing to the action:

```
//matrix.us9.list-manage.com/subscribe/post?u=c95e0a682d8937f12732b48d4&amp;id=863aebab0f
```

#### Footer

The footer on Docusaurus is far more sophisticated compared to the footer in https://polykey.io/.

image

image

We want to keep all the links in the polykey.io/docs/ except the `Company` column.

The `Company` column is replaced with the footer from `polykey.io`.

image

The logo in the polykey.io/docs should be removed. Replace it with the logo plus "POLYKEY" as in polykey.io.

Where it says `Polykey is a product of Matrix AI`, add an additional text underneath it with the copyright message. It should say the current year. See: https://github.com/facebook/docusaurus/discussions/6628

The entire footer colours should match it on https://polykey.io/.

#### Fonts

The https://polykey.io uses `Open Sans` and has a backup to `sans-serif`.

We are supposed to 3 kinds of fonts on our website:

1. Open Sans (backup: `sans-serif`) - used for headings and buttons and regular text and links
2. Roboto Mono (backup: `monospace`) - this is supposed to be used for code blocks

Buttons are always fully capital bold open sans.

Headings are usually bold open sans.

Smaller headings are usually semi-bold open sans.

The TTF versions of the fonts are available here: https://github.com/MatrixAI/Polykey-Desktop/tree/staging/src/renderer/fonts. Copy the 2 font directories to this project and put in the right place (static?).

#### Colours

We have background tones, content tones, accent colours and semantic colours.

These are the colours to be used:

image

```
// Grey
grey1: '#FFFFFF',
grey2: '#FDFEFF',
grey3: '#FAFAFA',
grey4: '#E7E9F2',
grey5: '#E8F0F0',
grey6: '#EEEFF2',

// Primary
primary1: '#116466',
primary2: '#289295',
primary3: '#7BBBBC',
primary4: '#D1E8E2',

// Semantic
semantic1: '#F15446',

// Content tones
content1: '#1F211B',
content2: '#7F88A5',
content3: '#C2C9DF',
content4: '#E4E9F2',
```

The content colours are used for text.

The primary colours are used for buttons and headings.

The semantic colour is used for errors.

The grey colours are used for background.

In the docusaurus website, it uses the Infima colours. Which uses 7 shades of colours per category.

https://infima.dev/docs/utilities/colors

Map the greys to the secondary colours. These will be the "background kind of colors".

Map the primary colours to the primary colours. These are the headings and links.

Map the content colours to the content colours. These are the text.

See this issue: https://github.com/facebook/docusaurus/issues/3955

This may require some fiddling to get right.

### Tasks

1. [x] Fix the side bar so we can use a custom layout
2. [ ] Bring in the Polykey theme as a custom theme

### PR DESCRIPTION
**Pull request by developerdesigner18**
### Description

Create-Custom-Navbar-Component For Document-site

Fixes #6 

### Tasks

- [ ] 1. Create NavBar Component
- [ ] 2. Style it the same as `polykey.io`
- [ ] 3. Create the Navbar links that needs to derive its configuration from `docusaurus.config.js` in the `config.themeConfig.navbar.items` array.

### Final checklist

* [ ] Domain specific tests
* [ ] Full tests
* [ ] Updated inline-comment documentation
* [ ] Lint fixed
* [ ] Squash and rebased
* [ ] Sanity check the final build

### CODE REVIEW

File: src/css/custom.css [hunk 1]
```diff
@@ -0,0 +1,164 @@
...
+/* ---------------------------Footer-------------------------- */
+@media (max-width: 768px) {
+    .header_menu {
+        display: none;
+    }
+
+    .header_mobile_menu {
+        display: block;
+    }
+}
```
Comment (CMCDragonkai) [inline src/css/custom.css:164]: This requires an empty newline. All files must have an empty newline.

File: src/theme/Navbar/index.js [hunk 2]
```diff
@@ -0,0 +1,106 @@
+import React from "react";
```
Comment (CMCDragonkai) [inline src/theme/Navbar/index.js:1]: Hey can you use `index.tsx` not `index.js`?

**Comment by CMCDragonkai** [comment]
There was all these subcomponents that are available for the NavBar, is overwriting the entire NavBar the right idea?

image

**Comment by CMCDragonkai** [comment]
In the future, please indicate the PR fixes the issue. But I've already done it above.

**Comment by CMCDragonkai** [comment]
Hi @developerdesigner18 in the future, please write your commit as `feat: ...` or `fix: ...`.

File: src/theme/Navbar/index.js [hunk 3]
```diff
@@ -0,0 +1,106 @@
...
+import NavbarLayout from "@theme/Navbar/Layout";
+import NavbarContent from "@theme/Navbar/Content";
+import useBaseUrl from "@docusaurus/useBaseUrl";
+import Logo from "../../../images/polykey-logo-light.svg";
```
Comment (CMCDragonkai) [inline src/theme/Navbar/index.js:5]: These components aren't being used.

File: docusaurus.config.js [hunk 4]
```diff
@@ -21,161 +21,161 @@ const remarkImageSrcWithRequire = (options) => {
...
   plugins: [
     [
-      '@docusaurus/plugin-content-docs',
+      "@docusaurus/plugin-content-docs",
       /** @type {import('@docusaurus/plugin-content-docs').Options} */
       ({
-        path: 'docs',
-        routeBasePath: '/',
-        sidebarPath: require.resolve('./sidebars.js'),
+        path: "docs",
```
Comment (CMCDragonkai) [inline docusaurus.config.js:44]: Hey @developerdesigner18 why did you do these styling changes?

Comment (CMCDragonkai) [reply docusaurus.config.js:44]: I'm bringing in my company's eslinting rules to avoid inconsistency.

File: src/theme/Navbar/index.js [hunk 5]
```diff
@@ -0,0 +1,106 @@
...
+            <ul>
+              <li>
+                <a href="https://polykey.io/blog">Blog</a>
+              </li>
+              <li>
+                <a href="https://polykey.io/docs">Docs</a>
+              </li>
+              <li>
+                <a href="https://polykey.io/download">Download</a>
+              </li>
```
Comment (CMCDragonkai) [inline src/theme/Navbar/index.js:26]: These links should come from the `navbar.items` configuration. It shouldn't be hardcoded here.

File: src/css/custom.css (hunk 1 above)
Comment (CMCDragonkai) [reply src/css/custom.css]: I've added the newline.

**Comment by CMCDragonkai** [comment]
hey @developerdesigner18 the navbar does not look exactly the same. It is not seamless.

navbar_inconsistency

Do you see not only are the colours different, but the sizing is different. The entire thing is not seamless.

File: src/theme/Navbar/index.js (hunk 3 above)
Comment (CMCDragonkai) [reply src/theme/Navbar/index.js]: Don't leave components imported if they aren't being used.

File: src/theme/Navbar/index.tsx [hunk 6]
```diff
@@ -1,14 +1,107 @@
...
+                  }}
+                />
+                <button
+                  onClick={() => {
+                    console.log(email);
+                  }}
+                >
+                  Get Started
+                </button>
+              </li>
```
Comment (CMCDragonkai) [inline src/theme/Navbar/index.tsx:51]: This is supposed to be a `form` element.

image

Plus you need to have the form action so the form actually works. Right now this form element doesn't do anything!

Comment (CMCDragonkai) [reply src/theme/Navbar/index.tsx:51]: See `polykey.io` view the source, and copy the HTML form.

**Comment by CMCDragonkai** [comment]
Hey @developerdesigner18 I think there was some misunderstanding on the navbar menu for mobile view. **I want to keep the docusaurus menu**. Do not change it to the menu in `polykey.io`. Keep it the same.

ONLY make it come out from the **right** instead of the **left**.

Right now the menu is incorrect. **It must be the same as it was before with docusaurus**.

**Comment by CMCDragonkai** [comment]
We must be able to navigate the documentation like before.

**Comment by CMCDragonkai** [comment]
Also can you not just override the entire navbar component. Use the existing components here: https://github.com/MatrixAI/Polykey-Docs/pull/12#issuecomment-1311419124

And override **only** the ones necessary for the minimal template changes.

**Comment by CMCDragonkai** [comment]
Note to self, to push to the remote PR:

```sh
git fetch origin pull/12/head:feature-navbar
git checkout feature-navbar
git remote add developerdesigner git@github.com:developerdesigner18/Polykey-Docs.git
git push developerdesigner HEAD:Create-New-Navbar-Component
```

**Comment by CMCDragonkai** [comment]
In total these are the problems:

1. Components being imported not being used.
2. NavBar items should be derived from the docusaurus configuration rather than being hardcoded.
3. NavBar does not look seamless from polykey.io
4. The email form is not working, and doesn't have the `action` set.

**Comment by CMCDragonkai** [comment]
Can you combine your other PR into this? Don't create 2 PRs for these things. Let's see it all together...

**Comment by CMCDragonkai** [comment]
The navbar background colour is `#0e0e0e`. How did you end up choosing `#242526` as the bg colour for the navbar?

**Comment by CMCDragonkai** [comment]
BTW we use conventional commits. Please follow the same style in the future.

**Comment by CMCDragonkai** [comment]
Hey @developerdesigner18 I want all the fonts in the `static/fonts`. Don't just put 1 font. Put all of them. The entire directories from https://github.com/MatrixAI/Polykey-Desktop/tree/staging/src/renderer/fonts.

**Comment by CMCDragonkai** [comment]
Hey @developerdesigner18 why did you delete the `index.tsx` again? We are meant to be using `index.tsx`... not `index.js`.

I already changed this last time, why did you end up deleting mine again?

File: src/theme/Navbar/index.js [hunk 7]
```diff
@@ -0,0 +1,155 @@
...
+                      .then((res) => {
+                        console.log(res);
+                      })
+                      .catch((err) => {
+                        console.log(err);
+                      });
+                  }}
+                >
+                  Get Started
+                </button>
```
Comment (CMCDragonkai) [inline src/theme/Navbar/index.js:67]: This should be using `action` on the form.

File: src/theme/Navbar/index.tsx (hunk 6 above)
Comment (developerdesigner18) [reply src/theme/Navbar/index.tsx]: https://imgur.com/DDyyMPq

File: src/theme/Navbar/index.js (hunk 7 above)
Comment (developerdesigner18) [reply src/theme/Navbar/index.js:67]: https://imgur.com/DDyyMPq

**Comment by developerdesigner18** [comment]
> Hey @developerdesigner18 I want all the fonts in the `static/fonts`. Don't just put 1 font. Put all of them. The entire directories from https://github.com/MatrixAI/Polykey-Desktop/tree/staging/src/renderer/fonts.

I have put all the fonts from the location that you have attached here.

**Comment by developerdesigner18** [comment]
> The navbar background colour is `#0e0e0e`. How did you end up choosing `#242526` as the bg colour for the navbar?

This things is done

**Comment by developerdesigner18** [comment]
> > Hey @developerdesigner18 I want all the fonts in the `static/fonts`. Don't just put 1 font. Put all of them. The entire directories from https://github.com/MatrixAI/Polykey-Desktop/tree/staging/src/renderer/fonts.
> 
> I have put all the fonts from the location that you have attached here.

**Comment by developerdesigner18** [comment]
> In total these are the problems:
> 
> 1. Components being imported not being used.
> 2. NavBar items should be derived from the docusaurus configuration rather than being hardcoded.
> 3. NavBar does not look seamless from polykey.io
> 4. The email form is not working, and doesn't have the `action` set.
> 5. Rather than overriding the **entire** component in `src/theme/Navbar`, you should be overriding the smaller components and allowing the theme to put it together (this might help with problem 2., so you can derive the content more easily)
> 6. The mobile menu is incorrect. It should be how docusaurus does it, because **we must have the docusaurus navigation working**. We only want it to come out from the right.

>Issue That is fixed
>1.All the component is imported.
>2. Navbar Items is derived from the docusaurus_config_file.
>3. Now Navbar looks like the main site.
>4. Email is also working.
>5. I'm trying to override the smaller component instead of overriding whole the component.

**Comment by CMCDragonkai** [comment]
> > Hey @developerdesigner18 I want all the fonts in the `static/fonts`. Don't just put 1 font. Put all of them. The entire directories from https://github.com/MatrixAI/Polykey-Desktop/tree/staging/src/renderer/fonts.
> 
> I have put all the fonts from the location that you have attached here.

No that's not what I see:

image

**Comment by CMCDragonkai** [comment]
And still no tsx..?

**Comment by developerdesigner18** [comment]
> And still no tsx..?

>I'm working on this.

**Comment by CMCDragonkai** [comment]
We aren't using Nunito Sans, so you can remove that.

Open Sans should be used for the font of the content.

The roboto should be used for code. I already explained all this in the original issue #6.

**Comment by developerdesigner18** [comment]
> We aren't using Nunito Sans, so you can remove that.
> 
> Open Sans should be used for the font of the content.
> 
> The roboto should be used for code. I already explained all this in the original issue #6.

>Okay Got it!

**Comment by CMCDragonkai** [comment]
@developerdesigner18 there's an extra `static` directory inside `Roboto_Mono` directory. What's this extra directory for? I think you might have accidentally created it.

**Comment by CMCDragonkai** [comment]
The current branch cannot start.

```
[nix-shell:~/Projects/Polykey-Docs]$ npm run start

> start
> docusaurus start

[INFO] Starting the development server...
[ERROR] ValidationError: The footer must be either simple or multi-column, and not a mix of the two. See: https://docusaurus.io/docs/api/themes/configuration#footer-links
[INFO] Docusaurus version: 2.2.0
Node version: v16.15.0
```

**Comment by developerdesigner18** [comment]
> The current branch cannot start.
> 
> ```
> [nix-shell:~/Projects/Polykey-Docs]$ npm run start
> 
> > start
> > docusaurus start
> 
> [INFO] Starting the development server...
> [ERROR] ValidationError: The footer must be either simple or multi-column, and not a mix of the two. See: https://docusaurus.io/docs/api/themes/configuration#footer-links
> [INFO] Docusaurus version: 2.2.0
> Node version: v16.15.0
> ```

>This Error comes because I have made some changes in the config file. I have updated the code now once the code run then please uncomment the items from the config file.
>Becasue we are showing different footer for Mobile And Desktop view. And I'm not able to add more components in the footer items in the config file.

**Comment by CMCDragonkai** [comment]
@bmamouri please work off this PR first when you start on Polykey-Docs.

**Comment by CMCDragonkai** [comment]
@amydevs consider using GitLab LFS or GitHub LFS to store images and then reference them within the content as if they were normal images on the file system.

This avoids the complication of S3, while also potentially allows them to still be rendered in GitHub/GitLab markdown rendering.

Cloudflare won't care, as long as the when you clone/pull the repo, the LFS is activated to download those images.

**Comment by amydevs** [comment]
@bmamouri I've just merged a branch into main regarding fixing image urls, so am commenting here as a heads-up to rebase

**Comment by bmamouri** [comment]
Closing this PR because it is outdated and we have the new PR #17 includes the useful work of this PR. Thanks

**Pull request closed without merging**